### PR TITLE
DB-8098 add doNotFlatten hint for subquery

### PR DESCRIPTION
**Pull request by yxia92**

### CODE REVIEW

**Comment by yxia92** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by yxia92** [comment]
jenkins please retest branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-shared/src/main/java/com/splicemachine/db/shared/common/reference/SQLState.java [hunk 1]
```diff
@@ -1044,6 +1044,8 @@ The message identifier (ie. the value that matches a field in this class)
 	String LANG_INVALID_USE_OF_DEFAULT                                 = "42Y85";
 	String LANG_INVALID_SESSION_PROPERTY                               = "42Y86";
 	String LANG_INVALID_SESSION_PROPERTY_VALUE                         = "42Y87";
+	String LANG_INVALID_FORCED_NO_SUBQUERY_FLATTEN_VALUE               = "42Y88";
+	String LANG_INVALID_SUBQUYERY_PROPERTY                             = "42Y89";
```
Comment (arnaud-splice) [inline db-shared/src/main/java/com/splicemachine/db/shared/common/reference/SQLState.java]: typo: LANG_INVALID_SUBQU{Y,}ERY_PROPERTY

Comment (yxia92) [reply db-shared/src/main/java/com/splicemachine/db/shared/common/reference/SQLState.java]: Thanks @arnaud-splice I've fixed the typo.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SubqueryNode.java [hunk 2]
```diff
@@ -220,6 +222,31 @@ public void init(
...
+            String value=(String)properties.get(key);
+
+            switch(key){
+                case "doNotFlatten":
+                    try {
+                        hintNotFlatten = Boolean.parseBoolean(StringUtil.SQLToUpperCase(value));
+                    } catch (Exception exception) {
+                        throw StandardException.newException(SQLState.LANG_INVALID_FORCED_NO_SUBQUERY_FLATTEN_VALUE, value);
+                    }
+                    break;
```
Comment (changli6) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SubqueryNode.java]: This 'break' only jumps out of the switch part, I think it's better to jump out of the while loop also.

Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SubqueryNode.java]: Thanks @changli6 . Right now, subquery only supports one hint, in the future, if we have more hints to support, we don't want to break out the while loop.

**Pull request merged**